File: archive/image_processing_lib/src/image_noise.c
/**
 * @file image_noise.c
 * @brief Implementation of noise generation functions
 */

#include "../include/image_noise.h"
#include <stdlib.h>
#include <math.h>
#include <time.h>

/* Initialize random number generator (call once) */
static int rng_initialized = 0;

static void init_rng() {
    if (!rng_initialized) {
        srand((unsigned int)time(NULL));
        rng_initialized = 1;
    }
}

/* Generate random number in [0, 1] */
static double rand_uniform() {
    return (double)rand() / (double)RAND_MAX;
}

/* Generate Gaussian random number using Box-Muller transform */
static double rand_gaussian(double mean, double stddev) {
    static int has_spare = 0;
    static double spare;
    
    if (has_spare) {
        has_spare = 0;
        return mean + stddev * spare;
    }
    
    has_spare = 1;
    double u, v, s;
    do {
        u = rand_uniform() * 2.0 - 1.0;
        v = rand_uniform() * 2.0 - 1.0;
        s = u * u + v * v;
    } while (s >= 1.0 || s == 0.0);
    
    s = sqrt(-2.0 * log(s) / s);
    spare = v * s;
    
    return mean + stddev * u * s;
}

int add_gaussian_noise(const BMPImage *input, BMPImage *output, double mean, double stddev) {
    int i;
    int temp;
    
    init_rng();
    
    /* Copy header and properties */
    if (bmp_copy(input, output) != 0) {
        return -1;
    }
    
    /* Add Gaussian noise to each pixel */
    for (i = 0; i < input->dataSize; i++) {
        temp = (int)(input->data[i] + rand_gaussian(mean, stddev));
        
        /* Clamp to valid range */
        if (temp > 255) temp = 255;
        if (temp < 0) temp = 0;
        
        output->data[i] = (unsigned char)temp;
    }
    
    return 0;
}

int add_salt_pepper_noise(const BMPImage *input, BMPImage *output, double probability) {
    int i;
    double r;
    
    init_rng();
    
    /* Copy header and properties */
    if (bmp_copy(input, output) != 0) {
        return -1;
    }
    
    /* Add salt-and-pepper noise */
    for (i = 0; i < input->dataSize; i++) {
        r = rand_uniform();
        
        if (r < probability / 2.0) {
            /* Pepper (black) */
            output->data[i] = 0;
        } else if (r < probability) {
            /* Salt (white) */
            output->data[i] = 255;
        } else {
            /* No noise */
            output->data[i] = input->data[i];
        }
    }
    
    return 0;
}

int add_uniform_noise(const BMPImage *input, BMPImage *output, int amplitude) {
    int i;
    int temp, noise;
    
    init_rng();
    
    /* Copy header and properties */
    if (bmp_copy(input, output) != 0) {
        return -1;
    }
    
    /* Add uniform noise to each pixel */
    for (i = 0; i < input->dataSize; i++) {
        noise = (int)((rand_uniform() * 2.0 - 1.0) * amplitude);
        temp = input->data[i] + noise;
        
        /* Clamp to valid range */
        if (temp > 255) temp = 255;
        if (temp < 0) temp = 0;
        
        output->data[i] = (unsigned char)temp;
    }
    
    return 0;
}

File: src/image_processing/image_noise.rs
use super::bmp_io::BmpImage;

// ---------------------------------------------------------------------------
// Gaussian noise
// ---------------------------------------------------------------------------

/// Add Gaussian noise to a grayscale image.
///
/// Uses the Box-Muller / Marsaglia polar method to generate normally-
/// distributed noise which is **added** to each existing pixel value.
/// `mean` is the noise centre (typically 0.0) and `stddev` is the standard
/// deviation.
///
/// A simple built-in PRNG is used so no external crate is needed.  Supply a
/// `seed` for reproducibility.
///
/// Translated from the C library `add_gaussian_noise` function.
#[must_use]
pub fn add_gaussian_noise(input: &BmpImage, mean: f64, stddev: f64, seed: u64) -> BmpImage {
    let mut output_data = input.data.clone();
    let mut rng = SimpleRng::new(seed);
    let mut has_spare = false;
    let mut spare: f64 = 0.0;

    for px in &mut output_data {
        let noise = if has_spare {
            has_spare = false;
            mean + stddev * spare
        } else {
            let (u, v, s) = loop {
                let u = rng.next_f64() * 2.0 - 1.0;
                let v = rng.next_f64() * 2.0 - 1.0;
                let s = u * u + v * v;
                if s > 0.0 && s < 1.0 {
                    break (u, v, s);
                }
            };
            let factor = (-2.0 * s.ln() / s).sqrt();
            spare = v * factor;
            has_spare = true;
            mean + stddev * u * factor
        };

        let val = (*px as f64 + noise).clamp(0.0, 255.0);
        *px = val as u8;
    }

    input.with_data(output_data)
}

// ---------------------------------------------------------------------------
// Salt-and-pepper noise
// ---------------------------------------------------------------------------

/// Add salt-and-pepper (impulse) noise to a grayscale image.
///
/// `probability` controls the total density of noise (0.0 = none, 1.0 = all).
/// Half the noisy pixels become 0 (pepper), the other half become 255 (salt).
///
/// Translated from the C library `add_salt_pepper_noise` function.
#[must_use]
pub fn add_salt_pepper_noise(input: &BmpImage, probability: f64, seed: u64) -> BmpImage {
    let mut output_data = input.data.clone();
    let mut rng = SimpleRng::new(seed);

    for px in &mut output_data {
        let r = rng.next_f64();
        if r < probability / 2.0 {
            *px = 0; // pepper (black)
        } else if r < probability {
            *px = 255; // salt (white)
        }
        // otherwise keep original value
    }

    input.with_data(output_data)
}

// ---------------------------------------------------------------------------
// Uniform noise
// ---------------------------------------------------------------------------

/// Add uniform noise to a grayscale image.
///
/// Each pixel is offset by a random value in the range `[-amplitude, +amplitude]`.
#[must_use]
pub fn add_uniform_noise(input: &BmpImage, amplitude: f32, seed: u64) -> BmpImage {
    let mut output_data = input.data.clone();
    let mut rng = SimpleRng::new(seed);

    for px in &mut output_data {
        let noise = rng.next_f32() * 2.0 * amplitude - amplitude;
        let val = (*px as f32 + noise).clamp(0.0, 255.0);
        *px = val as u8;
    }

    input.with_data(output_data)
}

// ---------------------------------------------------------------------------
// Minimal PRNG (no external dependency required)
// ---------------------------------------------------------------------------

/// A simple xorshift64-based PRNG so the noise functions do not need an
/// external random number crate.
struct SimpleRng {
    state: u64,
}

impl SimpleRng {
    fn new(seed: u64) -> Self {
        Self {
            state: if seed == 0 { 1 } else { seed },
        }
    }

    fn next_u64(&mut self) -> u64 {
        let mut x = self.state;
        x ^= x << 13;
        x ^= x >> 7;
        x ^= x << 17;
        self.state = x;
        x
    }

    fn next_u32(&mut self) -> u32 {
        (self.next_u64() >> 16) as u32
    }

    fn next_f32(&mut self) -> f32 {
        (self.next_u32() & 0x00FF_FFFF) as f32 / 16_777_216.0
    }

    fn next_f64(&mut self) -> f64 {
        (self.next_u64() >> 11) as f64 / (1u64 << 53) as f64
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn gaussian_noise_changes_image() {
        let mut img = BmpImage::new();
        img.width = 8;
        img.height = 8;
        img.data = vec![128u8; 64];
        let noisy = add_gaussian_noise(&img, 0.0, 20.0, 42);
        assert_ne!(noisy.data, img.data);
    }

    #[test]
    fn salt_pepper_zero_prob_is_identity() {
        let mut img = BmpImage::new();
        img.width = 8;
        img.height = 8;
        img.data = vec![128u8; 64];
        let noisy = add_salt_pepper_noise(&img, 0.0_f64, 42);
        assert_eq!(noisy.data, img.data);
    }

    #[test]
    fn uniform_noise_zero_amplitude_is_identity() {
        let mut img = BmpImage::new();
        img.width = 4;
        img.height = 4;
        img.data = vec![100u8; 16];
        let noisy = add_uniform_noise(&img, 0.0, 42);
        assert_eq!(noisy.data, img.data);
    }
}
